homework-assignments-1/solutions/task-4.ts

```diff
@@ -0,0 +1,200 @@
+class Product {
+  public name: string;
+  public price: number;
+  public stock: number;
+
+  constructor(name: string, price: number, stock: number) {
+    this.name = name;
+    this.price = price;
+    this.stock = stock;
+  }
+
+  addStock(quantity: number) {
+    this.stock += quantity;
+  }
+
+  reduceStock(quantity: number): void | never {
+    if (quantity > this.stock) {
+      throw new Error("Not sufficient stock");
+    }
+
+    this.stock -= quantity;
+  }
+}
+
+class Customer {
+  public name: string;
+  public email: string;
+  public cart: Product[];
+  public address: string;
+
+  constructor(name: string, email: string, address: string) {
+    this.name = name;
+    this.email = email;
+    this.address = address;
+
+    this.cart = [];
+  }
+
+  addToCart(product: Product) {
+    this.cart.push(product);
```

Comment:
dont be in rush to add product to cart. check whether there are any stocks available. if (product.getStocks() > 0)